lib/protoparser/prometheus/parser.go

```diff
@@ -727,3 +785,94 @@ var numericChars = [256]bool{
 	'E': true,
 	'.': true,
 }
+
+// Metadata contains HELP or TYPE message in prometheus exposition format.
+// See https://github.com/prometheus/docs/blob/e39897e4ee6e67d49d47204a34d120e3314e82f9/docs/instrumenting/exposition_formats.md#comments-help-text-and-type-information.
+//
+// For example:
+// # HELP alertmanager_alerts How many alerts by state.
+// # TYPE alertmanager_alerts gauge
+type Metadata struct {
+	Metric string
+	Type   uint32
+	Help   string
+}
+
+// MetadataRows contains parsed Prometheus metadata rows.
+type MetadataRows struct {
+	Rows []Metadata
+}
+
+// Reset resets ms.
+func (ms *MetadataRows) Reset() {
+	ms.Rows = ms.Rows[:0]
+}
+
+func parseMetadataLine(s string, dst []Metadata, errLogger func(s string)) []Metadata {
+	if len(s) > 0 && s[len(s)-1] == '\r' {
+		s = s[:len(s)-1]
+	}
+	s = skipLeadingWhitespace(s)
+
+	if len(s) < 2 || s[0] != '#' || s[1] != ' ' {
+		// Skip non-comment
+		return dst
+	}
+
+	parts := strings.SplitN(s, " ", 4)
+	if len(parts) < 3 {
+		if errLogger != nil {
+			errLogger(fmt.Sprintf("cannot unmarshal metadata line %q", s))
+		}
+		return dst
+	}
+
+	var md Metadata
```

Comment:
I think, it's better to add whole metadata in 1 go. Instead of joining it later at the caller side.